### PR TITLE
Storing new assignments

### PR DESCRIPTION
**Pull request by kalabukdima**

### CODE REVIEW

File: crates/network-scheduler/src/server.rs [hunk 1]
```diff
@@ -171,7 +172,11 @@ impl Server {
                 if current_epoch >= last_schedule_epoch + schedule_interval {
                     scheduler.schedule(current_epoch);
                     match scheduler.to_json() {
-                        Ok(state) => storage_client.save_scheduler(state).await,
+                        Ok(state) => {
```
Comment (kalabukdima) [inline crates/network-scheduler/src/server.rs:177]: This is only done once per 3 hours. Assignments should be uploaded on each change, or just in this function which is run every minute: https://github.com/subsquid/network-components/blob/639544acb102554e5a5d0ad805a1037e76ab7848/crates/network-scheduler/src/server.rs#L293

File: crates/network-scheduler/src/storage.rs [hunk 2]
```diff
@@ -301,6 +308,53 @@ impl S3Storage {
...
+        let mut encoder = GzEncoder::new(Vec::new(), Compression::best());
+        let _ = encoder.write_all(serde_json::to_vec(&assignment).unwrap().as_slice());
+        let compressed_bytes = encoder.finish().unwrap();
+        log::debug!("Saving assignment");
+        let mut hasher = Sha256::new();
+        hasher.update(compressed_bytes.as_slice());
+        let hash = hasher.finalize();
+        let network = Config::get().network.clone();
+        let dt = Utc::now();
+        let ts = dt.format("%Y%m%dT%H%M%S");
```
Comment (kalabukdima) [inline crates/network-scheduler/src/storage.rs]: May be shortened to `%FT%T`

File: crates/network-scheduler/src/storage.rs [hunk 3]
```diff
@@ -301,6 +308,53 @@ impl S3Storage {
...
+        let assignment = Assignment::new(&json, Config::get().cloudflare_storage_secret.clone());
+        let mut encoder = GzEncoder::new(Vec::new(), Compression::best());
+        let _ = encoder.write_all(serde_json::to_vec(&assignment).unwrap().as_slice());
+        let compressed_bytes = encoder.finish().unwrap();
+        log::debug!("Saving assignment");
+        let mut hasher = Sha256::new();
+        hasher.update(compressed_bytes.as_slice());
+        let hash = hasher.finalize();
+        let network = Config::get().network.clone();
+        let dt = Utc::now();
```
Comment (kalabukdima) [inline crates/network-scheduler/src/storage.rs]: Please avoid abbreviations

File: crates/network-scheduler/src/storage.rs [hunk 4]
```diff
@@ -301,6 +308,53 @@ impl S3Storage {
...
+        let ts = dt.format("%Y%m%dT%H%M%S");
+        let filename: String =  format!("assignments/{network}_{ts}_{hash:X}.json.gz");
```
Comment (kalabukdima) [inline crates/network-scheduler/src/storage.rs]: I think `assignments/{network}/{ts}_{hash}.json.gz` might be more convenient for browsing in the UI

File: crates/network-scheduler/src/storage.rs [hunk 5]
```diff
@@ -301,6 +308,53 @@ impl S3Storage {
         prometheus_metrics::s3_request();
     }
 
+    pub async fn save_assignment(&self, scheduler_state: &Vec<u8>) {
+        log::debug!("Encoding assignment");
+        let json: Value = serde_json::from_slice(scheduler_state).unwrap();
```
Comment (kalabukdima) [inline crates/network-scheduler/src/storage.rs]: This is a very heavy operation with a lot of memory allocations. Can the assignment be generated from the `scheduler_state` directly or even from the `chunks_summary`?

File: crates/network-scheduler/src/assignment.rs [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+use crate::signature::timed_hmac_now;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct Chunk {
+    pub id: String,
+    pub base_url: String,
+    pub files: HashMap<String, String>,
+    size: u64,
```
Comment (kalabukdima) [inline crates/network-scheduler/src/assignment.rs]: Let's call it `sizeBytes`

File: crates/network-scheduler/src/assignment.rs [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+#[derive(Serialize, Deserialize)]
+pub struct NetworkState {
+    pub(crate) network: String,
+    pub(crate) assignment: NetworkAssignment
+}
+
+impl Assignment {
+    fn add_chunk(&mut self, chunk: Chunk, dataset_id: String, dataset_url: String) {
+        match self.datasets.iter_mut().find(|dataset| dataset.id == dataset_id) {
```
Comment (kalabukdima) [inline crates/network-scheduler/src/assignment.rs:62]: I think it's easier to fill a BTreeMap and dump it into a vector after filling

File: crates/network-scheduler/src/storage.rs [hunk 8]
```diff
@@ -301,6 +308,53 @@ impl S3Storage {
...
+        
+        let _ = self
+            .client
+            .put_object()
+            .bucket(&self.config.scheduler_state_bucket)
+            .key(&filename)
+            .body(compressed_bytes.into())
+            .send()
+            .await
+            .map_err(|e| log::error!("Error saving assignment: {e:?}"));
```
Comment (kalabukdima) [inline crates/network-scheduler/src/storage.rs]: If an error occurs here, we should return from this function. It's easier if it returns a `Result` and the logging is done outside of it

File: crates/network-scheduler/src/assignment.rs [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+                break;
+            }
+            let mut filtered_chunks: Vec<Chunk> = Default::default();
+            for v in &u.chunks {
+                if idxs[0] < cursor {
+                    return None; // Malformed diffs
+                }
+                if idxs[0] == cursor {
+                    filtered_chunks.push(v.clone());
+                    idxs.remove(0);
```
Comment (kalabukdima) [inline crates/network-scheduler/src/assignment.rs:103]: Why? It's unnecessarily O(n^2). I would do it like this:
```rust
let flat_list: Vec<(&Dataset, &Chunk)> = self.datasets
    .iter()
    .map(|dataset| {
        dataset
            .chunks
            .iter()
            .map(|chunk| (dataset, chunk))
    })
    .flatten()
    .collect();
let mut result: HashMap<String, Dataset> = Default::default();
for index in delta_decode(&local_assignment.chunks_deltas) {
    let (dataset, chunk) = flat_list.get(index as usize)?;
    result.entry(dataset.id).or_insert_with(|| Dataset {
        id: dataset.id.clone(),
        base_url: dataset.base_url.clone(),
        chunks: vec![],
    }).chunks.push((*chunk).clone());
}
Some(result.into_values().collect())
```

BTW, this method should consume `self` and try to reuse the allocated memory

File: crates/network-scheduler/src/assignment.rs [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+                    chunk_map.insert(chunk.id.clone(), idx);
+                    idx += 1;
+                }
+            };
+            self.chunk_map = Some(chunk_map);
+        };
+        match self.chunk_map.as_ref().unwrap().get(&chunk_id) {
+            Some(idx) =>Some(idx.clone()),
+            None => None
+        }
```
Comment (kalabukdima) [inline crates/network-scheduler/src/assignment.rs]: There are quite a lot of useful clippy warnings

File: crates/network-scheduler/src/storage.rs (hunk 2 above)
Comment (denisbsu) [reply crates/network-scheduler/src/storage.rs]: fixed

File: crates/network-scheduler/src/storage.rs (hunk 3 above)
Comment (denisbsu) [reply crates/network-scheduler/src/storage.rs]: fixed

File: crates/network-scheduler/src/storage.rs (hunk 4 above)
Comment (denisbsu) [reply crates/network-scheduler/src/storage.rs]: fixed

File: crates/network-scheduler/src/assignment.rs (hunk 6 above)
Comment (denisbsu) [reply crates/network-scheduler/src/assignment.rs]: fixed

File: crates/network-scheduler/src/storage.rs (hunk 8 above)
Comment (denisbsu) [reply crates/network-scheduler/src/storage.rs]: fixed

File: crates/network-scheduler/src/assignment.rs (hunk 9 above)
Comment (denisbsu) [reply crates/network-scheduler/src/assignment.rs:103]: fixed

File: crates/network-scheduler/src/assignment.rs (hunk 10 above)
Comment (denisbsu) [reply crates/network-scheduler/src/assignment.rs]: fixed

File: crates/network-scheduler/src/storage.rs (hunk 5 above)
Comment (denisbsu) [reply crates/network-scheduler/src/storage.rs]: done

File: crates/network-scheduler/src/server.rs (hunk 1 above)
Comment (denisbsu) [reply crates/network-scheduler/src/server.rs:177]: moved

File: crates/network-scheduler/src/storage.rs [hunk 11]
```diff
@@ -301,6 +307,59 @@ impl S3Storage {
...
+        let _ = encoder.write_all(serde_json::to_vec(&assignment).unwrap().as_slice());
+        let compressed_bytes = encoder.finish().unwrap();
+        log::debug!("Saving assignment");
+        let mut hasher = Sha256::new();
+        hasher.update(compressed_bytes.as_slice());
+        let hash = hasher.finalize();
+        let network = Config::get().network.clone();
+        let current_time = Utc::now();
+        let timestamp = current_time.format("%FT%T");
+        let filename: String =  format!("assignments/{network}/{timestamp}_{hash:X}.json.gz");
```
Comment (kalabukdima) [inline crates/network-scheduler/src/storage.rs:322]: How long is this hash? Let's trim it to 8-10 bytes max

**Pull request merged**